### PR TITLE
[TESTMERGE ONLY] Unrolls keyLoop in SSInput.

### PR DESCRIPTION
**Pull request by Iamgoofball**
Let's see if we can eek out some performance by dropping the proc call overhead from a hot loop.

### CODE REVIEW

File: code/controllers/subsystem/input.dm [hunk 1]
```diff
@@ -36,4 +36,34 @@ SUBSYSTEM_DEF(input)
...
 	for(var/mob/user as anything in GLOB.keyloop_list)
-		user.focus?.keyLoop(user.client)
+		if(user.focus)
+			var/movement_dir = NONE
+			for(var/_key in user.client?.keys_held)
+				movement_dir = movement_dir | user.movement_keys[_key]
+			if(user?.next_move_dir_add)
+				movement_dir |= user.client.next_move_dir_add
+			if(user?.next_move_dir_sub)
+				movement_dir &= ~user.client.next_move_dir_sub
```
Comment (LemonInTheDark) [inline code/controllers/subsystem/input.dm:46]: I'm reasonably sure all these var accesses outweigh any benefit of unrolling the loop. A proc with no args isn't actually that expensive to call. This also means that any runtimes in keyLoop break ssinput, which is worrying.

Shouldn't need to run this on live to profile it, have you checked yourself?

Comment (Iamgoofball) [reply code/controllers/subsystem/input.dm:46]: The proc had an arg, it fed in the client.

Comment (LemonInTheDark) [reply code/controllers/subsystem/input.dm:46]: oh that's true. I wonder if it's still faster in just the base case of input firing. Would be good to see some clean profiles

File: code/controllers/subsystem/input.dm [hunk 2]
```diff
@@ -36,4 +36,38 @@ SUBSYSTEM_DEF(input)
...
+			// Sanity checks in case you hold left and right and up to make sure you only go up
+			if((movement_dir & NORTH) && (movement_dir & SOUTH))
+				movement_dir &= ~(NORTH|SOUTH)
+			if((movement_dir & EAST) && (movement_dir & WEST))
+				movement_dir &= ~(EAST|WEST)
+
+			if(user.client && movement_dir) //If we're not moving, don't compensate, as byond will auto-fill dir otherwise
+				movement_dir = turn(movement_dir, -dir2angle(user.client.dir)) //By doing this we ensure that our input direction is offset by the client (camera) direction
+
+			if(user.client?.movement_locked && user.focus)
```
Comment (Kylerace) [inline code/controllers/subsystem/input.dm:56]: why are you checking for focus here

File: code/controllers/subsystem/input.dm [hunk 3]
```diff
@@ -36,4 +36,38 @@ SUBSYSTEM_DEF(input)
...
+				if(isliving(user.focus))
```
Comment (Kylerace) [inline code/controllers/subsystem/input.dm:57]: these istype checks might actually make this inlining more expensive than calling a proc with overrides

**Review by Kylerace — CHANGES_REQUESTED** [review]

**Comment by LemonInTheDark** [comment]
I got bored and profiled this myself
```
Proc Name   Self CPU    Total CPU    Real Time    Overtime   Calls
------------------------------------------------------------------------------------------
/datum/controller/master/proc/RunQueue	            12.445 1747.137 1747.307 0.000 49614
/datum/controller/subsystem/proc/ignite	            2.340 1730.815 1733.861 0.001 1071256
/datum/controller/subsystem/input/fire	            161.857 1398.601 1398.743 0.007 49524
/datum/controller/subsystem/input/proc/old_fire	    171.311 647.264 667.012 0.014 107421843
/datum/controller/subsystem/input/proc/new_fire	    434.528 590.384 610.326 0.030 107422427
/atom/movable/keyLoop	                            309.616 475.951 495.500 0.024 107421843
/client/Move	                                    321.289 322.107 362.722 0.024 214844270
```
This was done with this commit <https://github.com/LemonInTheDark/tgstation/commit/f5e49136efa42e1bc0f3b86c0fb1917981232f94>

I've done some averaging and delta taking and such here https://www.desmos.com/calculator/dquzpkavzr
with the goal of figuring out just how big a savings we're looking at.
With the setDir unrolling this saves about 8.7% of the cost of standing still with a client. That cost is microscopic as it is though, it would take about 26 minutes to save a second of cpu time at our low highpop, so 60 players.

Here's the same logic and behavior, without setDir. It seems like kyler was right, unrolling setDir actually just increased costs, as without we save a second at 25 minutes, saving 9.1% overall. 
https://www.desmos.com/calculator/3cjtbz7ay7

Personally given how small the actual cost is here, I think the protection against runtimes breaking all input is much more important then the small saving. Especially because I was standing still for these tests. Any actual movement quickly makes these numbers semi redundant. Not sure it's really worth it.

**Comment by ZeWaka** [comment]
pr to goon when

**Comment by Mothblocks** [comment]
> That cost is microscopic as it is though, it would take about 26 minutes to save a second of cpu time at our low highpop, so 60 players.

**Pull request closed without merging**